tests/py/test_routes.py

```diff
@@ -26,34 +24,77 @@ def test_associate_and_delete_valid_card(self):
         assert len(cards) == 1
         assert self.roman.get_credit_card_error() == ''
 
-        self.hit('roman', 'delete', 'braintree-cc', cards[0].token)
+        address = cards[0].token
+        self.hit('roman', 'delete', 'braintree-cc', address)
 
         customer = self.roman.get_braintree_account()
         assert len(customer.credit_cards) == 0
 
         roman = P('roman')
-        assert roman.get_credit_card_error() == 'invalidated'
+        assert roman.get_credit_card_error() is None
+        assert self.db.one('select is_deleted from exchange_routes '
+                           'where address=%s', (address,))
         assert roman.braintree_customer_id
 
     def test_associate_invalid_card(self):
         self.hit('roman', 'associate', 'braintree-cc', 'an-invalid-nonce', expected=400)
         assert self.roman.get_credit_card_error() is None
 
-    @mock.patch.object(Participant, 'send_email')
-    def test_associate_paypal(self, mailer):
-        mailer.return_value = 1 # Email successfully sent
-        self.roman.add_email('roman@gmail.com')
-        self.db.run("UPDATE emails SET verified=true WHERE address='roman@gmail.com'")
+
+    def add_and_verify_email(self, participant, email_address):
+        participant.add_email(email_address)
+        self.db.run("UPDATE emails SET verified=true WHERE address=%s", (email_address,))
+
+    def test_associate_and_delete_valid_paypal(self):
+        self.add_and_verify_email(self.roman, 'roman@gmail.com')
+
         self.hit('roman', 'associate', 'paypal', 'roman@gmail.com')
         assert ExchangeRoute.from_network(self.roman, 'paypal')
         assert self.roman.has_payout_route
 
+        self.hit('roman', 'delete', 'paypal', 'roman@gmail.com')
+        assert not self.roman.has_payout_route
+
+    def test_delete_paypal_with_exchanges(self):
+        self.add_and_verify_email(self.roman, 'roman@gmail.com')
+        self.hit('roman', 'associate', 'paypal', 'roman@gmail.com')
+
+        self.make_exchange('paypal', 10, 0, self.roman)
+        self.hit('roman', 'delete', 'paypal', 'roman@gmail.com')
+
+        assert not self.roman.has_payout_route
+
+    def test_add_new_paypal_address(self):
+        self.add_and_verify_email(self.roman, 'roman@gmail.com')
+        self.add_and_verify_email(self.roman, 'candle@gmail.com')
+
+        self.hit('roman', 'associate', 'paypal', 'roman@gmail.com')
+        self.hit('roman', 'delete', 'paypal', 'roman@gmail.com')
+        self.hit('roman', 'associate', 'paypal', 'candle@gmail.com')
```

Comment:
could you extract above 5 lines to separate method, and reuse in test case below?